### PR TITLE
dag driver shutdown

### PR DESCRIPTION
**Pull request by zjma**

### CODE REVIEW

File: consensus/src/epoch_manager.rs [hunk 1]
```diff
@@ -759,8 +761,9 @@ impl EpochManager {
             self.time_service.clone(),
             genesis_block_id,
         );
-
-        tokio::spawn(dag_driver.start());
+        let (close_tx, close_rx) = oneshot::channel();
+        self.dag_driver_close_tx = Some(close_tx);
+        tokio::spawn(dag_driver.start(close_rx));
```
Comment (sasha8) [inline consensus/src/epoch_manager.rs:766]: We should call it from shutdown_current_processor

File: consensus/src/dag/dag_driver.rs
Comment (sasha8) [inline consensus/src/dag/dag_driver.rs:1]: We should also shutdown the worker in ReliableBroadcast.

**Comment by zjma** [comment]
This change makes `consensus_fault_tolerance::test_no_failures` failing again....

File: consensus/src/epoch_manager.rs (hunk 1 above)
Comment (zjma) [reply consensus/src/epoch_manager.rs:766]: Fixed.

File: consensus/src/dag/dag_driver.rs
Comment (zjma) [reply consensus/src/dag/dag_driver.rs:1]: Fixed.

File: consensus/src/dag/dag_driver.rs [hunk 2]
```diff
@@ -224,10 +232,23 @@ impl DagDriver {
...
+                close_req = close_rx.select_next_some() => {
+                    if let Some(close_tx) = self.rb_close_tx.take() {
+                        let (ack_tx, ack_rx) = oneshot::channel();
+                        close_tx.send(ack_tx).expect("[DagDriver] failed to drop rb");
+                        ack_rx.await.expect("[DagDriver] failed to drop rb");
+                    }
+                    if let Ok(ack_sender) = close_req {
+                        ack_sender.send(()).expect("[DagDriver] Fail to ack shutdown");
+                    }
+                    break;
```
Comment (zjma) [inline consensus/src/dag/dag_driver.rs:249]: synchronized wait for the ack from rb. Is that what we want? @sasha8

File: consensus/src/dag/dag_driver.rs [hunk 3]
```diff
@@ -108,6 +112,7 @@ impl DagDriver {
             ),
             bullshark,
             rb_tx,
+            rb_close_tx: Some(rb_close_tx),
```
Comment (sasha8) [inline consensus/src/dag/dag_driver.rs:115]: Does it have to be Some?

File: consensus/src/dag/dag_driver.rs (hunk 2 above)
Comment (sasha8) [reply consensus/src/dag/dag_driver.rs:249]: Yes.

**Review by sasha8 — APPROVED** [review]

File: consensus/src/dag/dag_driver.rs (hunk 3 above)
Comment (zjma) [reply consensus/src/dag/dag_driver.rs]: Surely it doesn't. Fixed.

**Pull request merged**